core/src/main/java/org/elasticsearch/index/query/TermsQueryBuilder.java

```diff
@@ -185,43 +192,108 @@ public String fieldName() {
     }
 
     public List<Object> values() {
-        return convertToStringListIfBytesRefList(this.values);
+        return convertBack(this.values);
     }
 
     public TermsLookup termsLookup() {
         return this.termsLookup;
     }
 
-    /**
-     * Same as {@link #convertToBytesRefIfString} but on Iterable.
-     * @param objs the Iterable of input object
-     * @return the same input or a list of {@link BytesRef} representation if input was a list of type string
-     */
-    private static List<Object> convertToBytesRefListIfStringList(Iterable<?> objs) {
-        if (objs == null) {
-            return null;
-        }
-        List<Object> newObjs = new ArrayList<>();
-        for (Object obj : objs) {
-            newObjs.add(convertToBytesRefIfString(obj));
+    private static final Set<Class<? extends Number>> INTEGER_TYPES = new HashSet<>(
+            Arrays.asList(Byte.class, Short.class, Integer.class, Long.class));
+    private static final Set<Class<?>> STRING_TYPES = new HashSet<>(
+            Arrays.asList(BytesRef.class, String.class));
+
+    private static List<?> convert(Iterable<?> values) {
```

Comment:
can we get some java docs on this just to make sure we don't loose the info why we did all this?